### PR TITLE
Block nonlinear forms and incompressible hyperelastic example [mixednonlin-dev]

### PR DESCRIPTION
**Pull request by jamiebramwell**
This adds new BlockNonlinearForm and BlockNonlinearFormIntegrator classes that allow computation of the action and gradient of nonlinear operators on block state vectors. It also adds a new example which models static finite deformation of an incompressible neo-Hookean (rubber-like) solid.  

TODO:
- [x] Update `doc/CodeDocumentation`
- [x] Update `CHANGELOG`

### CODE REVIEW

**Comment by jamiebramwell** [comment]
I'm trying to get away from being lazy and doing a system-wide direct solve, so I had to modify the built in Newton solver to allow for recalculation of the Jacobian preconditioner at every iteration. All of these solver interactions are getting a little complicated, so let me know if you think something else would work better.

File: examples/makefile [hunk 1]
```diff
@@ -23,7 +23,7 @@ MFEM_LIB_FILE = mfem_is_not_built
 
 SEQ_EXAMPLES = ex1 ex2 ex3 ex4 ex5 ex6 ex7 ex8 ex9 ex10 ex14 ex15 ex16 ex17
 PAR_EXAMPLES = ex1p ex2p ex3p ex4p ex5p ex6p ex7p ex8p ex9p ex10p ex11p ex12p\
- ex13p ex14p ex15p ex16p ex17p
+ ex13p ex14p ex15p ex16p ex17p ex18p
```
Comment (tzanio) [inline examples/makefile]: ex19p?

File: examples/makefile [hunk 2]
```diff
@@ -82,7 +82,7 @@ clean-build:
 clean-exec:
 	@rm -f refined.mesh displaced.mesh mesh.* ex5.mesh
 	@rm -rf Example5* Example9* Example15* Example16*
-	@rm -f sphere_refined.* sol.* sol_u.* sol_p.*
+	@rm -f sphere_refined.* sol.* sol_u.* sol_p.* deformation.* pressure.*
```
Comment (tzanio) [inline examples/makefile]: Add these also to `.gitignore`

File: makefile [hunk 3]
```diff
@@ -182,8 +182,8 @@ ifeq ($(MFEM_USE_OPENMP),YES)
 endif
 
 # List of MFEM dependencies, processed below
-MFEM_DEPENDENCIES = METIS LIBUNWIND SIDRE LAPACK OPENMP SUNDIALS MESQUITE\
- SUITESPARSE SUPERLU STRUMPACK GECKO GNUTLS NETCDF PETSC MPFR
+MFEM_DEPENDENCIES = SUPERLU METIS LIBUNWIND SIDRE LAPACK OPENMP SUNDIALS MESQUITE\
```
Comment (tzanio) [inline makefile]: Any reason the order was changed here?

File: fem/pnonlinearform.hpp [hunk 4]
```diff
@@ -57,7 +57,37 @@ class ParNonlinearForm : public NonlinearForm
    /// Set the operator type id for the parallel gradient matrix/operator.
    void SetGradientType(Operator::Type tid) { pGrad.SetType(tid); }
 
-   virtual ~ParNonlinearForm() { }
+   virtual ~ParNonlinearForm() { };
```
Comment (tzanio) [inline fem/pnonlinearform.hpp]: `{ };` -> `{ }`

File: makefile (hunk 3 above)
Comment (jamiebramwell) [reply makefile]: Yes. At least on the lab linux clusters using intel and clang compilers, superlu was missing some metis calls in the linking step due to it being last in the linking list.

File: examples/makefile (hunk 2 above)
Comment (jamiebramwell) [reply examples/makefile]: done.

File: fem/pnonlinearform.hpp (hunk 4 above)
Comment (jamiebramwell) [reply fem/pnonlinearform.hpp]: done.

File: fem/nonlininteg.hpp [hunk 5]
```diff
@@ -44,6 +44,38 @@ class NonlinearFormIntegrator
    virtual ~NonlinearFormIntegrator() { }
 };
 
+class BlockNonlinearFormIntegrator
```
Comment (jdahm) [inline fem/nonlininteg.hpp]: I apologize if I just haven't seen this yet, but is there a specific reason we need BlockNonlinearFormIntegrator and cannot just use NonlinearFormIntegrator?

While we're making that change, I think we could also add methods that deal with different test and trial spaces, since they'll be necessary when you have non-square blocks (... I have HDG in mind here).

Comment (jdahm) [reply fem/nonlininteg.hpp]: Oh, never mind I spoke too soon. I see the `Array<Vector *>` arguments now. I still think it would be nice to eventually have support for different test and trial spaces, however (second point above). Great work!

Comment (jamiebramwell) [reply fem/nonlininteg.hpp]: Yep! The idea here is that one could have terms involving variables from all of the function spaces. In the incompressible elasticity example, this includes terms that involve the product of the pressure and the deformation gradient.

These mixed formulations are possible, but it would involve a new nonlinear form type class that has two vectors of FE spaces: one for the test spaces and another for the trial ones. It shouldn't be too hard to add if this structure is followed.

File: fem/nonlinearform.cpp [hunk 6]
```diff
@@ -154,4 +154,464 @@ NonlinearForm::~NonlinearForm()
...
+            for (int s = 0; s < fes.Size(); s++)
+            {
+               fe[s] = fes[s]->GetFE(i);
+               fes[s]->GetElementVDofs(i, *vdofs[s]);
+               xs[s]->GetSubVector(*vdofs[s], *el_x[s]);
+            }
+
+            for (int k = 0; k < dfi.Size(); k++)
+            {
+               ffi[k]->AssembleElementGrad(fe, *T,
```
Comment (jdahm) [inline fem/nonlinearform.cpp]: Should this be a call to AssembleFaceGrad? I can add a commit later that adds interior face integrator support as well to bring it to parity with BilinearForm.

Comment (jamiebramwell) [reply fem/nonlinearform.cpp]: Yep! Now that the hydro-dev is merged, I'll change this to look more like that new code.

Comment (jamiebramwell) [reply fem/nonlinearform.cpp]: I added face function stubs to the generic block nonlinear integrator class. Let me know if you think something else should be added.

**Comment by jamiebramwell** [comment]
Both ex19 and ex19p now work with iterative solvers. I know I am probably over-solving in my Jacobian preconditioner, but this is the only way I could get it to work. I am open to suggestions from people who are smarter about these things.

Assuming the solver is OK, my next step is going back and adding BlockNonlinearForms to ex18 unless @tzanio thinks it should be a separate pull request.

**Comment by tzanio** [comment]
There seem to be an issue with convergence of the parallel preconditioner, so the Travis error in https://travis-ci.org/mfem/mfem/jobs/310737776 is correct. @jamiebramwell, maybe we can discuss the implementation and I can take a look?

**Comment by jamiebramwell** [comment]
All help is appreciated! I just noticed that it converges without issue when compiled in debug mode, but does not converge in optimized. I'll try to dig into the a little deeper.

**Comment by v-dobrev** [comment]
It is possible that the test run for ex19p is large and either does not fit in memory or just takes a lot of time.

File: fem/nonlininteg.cpp [hunk 7]
```diff
@@ -56,6 +56,58 @@ double NonlinearFormIntegrator::GetElementEnergy(
    return 0.0;
 }
 
+void BlockNonlinearFormIntegrator::AssembleElementVector(
```
Comment (acfisher) [inline fem/nonlininteg.cpp]: Why not make these methods pure virtual so that you don't have to write theses errors??

File: fem/nonlininteg.cpp [hunk 8]
```diff
@@ -385,4 +437,226 @@ void HyperelasticNLFIntegrator::AssembleElementGrad(const FiniteElement &el,
...
+      // u,p and p,u blocks
+      for (int i_p = 0; i_p < dof_p; i_p++)
+      {
+         for (int j_u = 0; j_u < dof_u; j_u++)
+         {
+            for (int dim_u = 0; dim_u < dim; dim_u++)
+            {
+               for (int l=0; l<dim; l++)
+               {
+                  (*elmats(1,0))(i_p, j_u + dof_u * dim_u) += dJ * FinvT(dim_u,l) * DS_u(j_u,
```
Comment (acfisher) [inline fem/nonlininteg.cpp]: You can save a little computation here by computing dJ*FinvT*DS_u*W once and using it for both elmats(0,1) and elmats(1,0).

File: linalg/superlu.cpp [hunk 9]
```diff
@@ -163,6 +163,158 @@ SuperLURowLocMatrix::SuperLURowLocMatrix( const HypreParMatrix & hypParMat )
...
+
+   // Fill csr arrays
+   for (int i=0; i<numsysrows; i++)
+   {
+      for (int j=0; j<numsyscols; j++)
+      {
+         if (!ops.IsZeroBlock(i,j))
+         {
+            Operator &op = ops.GetBlock(i,j);
+            parcsrmats(i,j) = (hypre_ParCSRMatrix *)dynamic_cast<HypreParMatrix&>(op);
```
Comment (acfisher) [inline linalg/superlu.cpp]: This dynamic cast looks a little fishy casting from HypreParMatrix and promoting to ParCSR, what if it's not a ParCSR?

File: fem/nonlininteg.cpp (hunk 7 above)
Comment (jamiebramwell) [reply fem/nonlininteg.cpp]: I followed the pattern of the original nonlinearformintegrator class. I agree that pure virtual should catch if these methods are overloaded at compile-time. Any thoughts @tzanio and @v-dobrev ?

File: fem/nonlininteg.cpp (hunk 8 above)
Comment (jamiebramwell) [reply fem/nonlininteg.cpp]: Ah yes! I remember I was going to do some optimizations in these compute-heavy loops and I totally forgot!
- [ ] Optimize gradient calculation

File: fem/nonlininteg.cpp (hunk 7 above)
Comment (v-dobrev) [reply fem/nonlininteg.cpp]: The base integrator class `NonlinearFormIntegrator` defines two types of virtual methods that a derived class can implement: integration over the elements, and/or integration over the faces. With the default implementation in the base class, the derived classes need to implement only the one that makes sense for them.

File: linalg/superlu.cpp (hunk 9 above)
Comment (jamiebramwell) [reply linalg/superlu.cpp]: That's a good question. I was honestly just following the pattern of the [`SuperLURowLocMatrix(const HypreParMatrix & hypParMat)`](https://github.com/mfem/mfem/blob/a6f1582bf8809cefa0d3676c8cbf51e2172cbbab/linalg/superlu.cpp#L124) constructor. I'll take a closer look at this.
- [ ] Examine ParCSR casting issue

**Comment by jamiebramwell** [comment]
Running ex19p through memcheck (per @acfisher's suggestion) yields errors and convergence failure even in debug mode. I'm going through the memcheck output now, but all signs point to a memory issue.

**Comment by jamiebramwell** [comment]
It looks like @v-dobrev fixed my memory/trueOffset issue while I was waiting for an account. Thanks for the help! I'll add a corresponding documentation pull request now.

File: makefile [hunk 10]
```diff
@@ -196,8 +196,7 @@ ifeq ($(MFEM_USE_OPENMP),YES)
 endif
 
 # List of MFEM dependencies, that require the *_LIB variable to be non-empty
-MFEM_REQ_LIB_DEPS = METIS SIDRE LAPACK SUNDIALS MESQUITE SUITESPARSE SUPERLU\
- STRUMPACK GECKO GNUTLS NETCDF PETSC MPFR
+MFEM_REQ_LIB_DEPS = SUPERLU METIS SIDRE LAPACK SUNDIALS MESQUITE SUITESPARSE STRUMPACK GECKO GNUTLS NETCDF PETSC MPFR
```
Comment (tzanio) [inline makefile]: Why this change?

**Comment by tzanio** [comment]
Listed some TODO items in PR description.

File: makefile (hunk 10 above)
Comment (jamiebramwell) [reply makefile]: SUPERLU uses METIS calls that are not used elsewhere. At least on LC, the link step requires that the dependencies be listed in order or else the extra functions are not included and the build fails.

**Comment by tzanio** [comment]
I get an error with `mpirun -np 4 ex19p` in a `make pdebug` build. 

I don't see it in 2cc790a844579d995. 

Any ideas what can be wrong?

**Comment by v-dobrev** [comment]
I don't get any errors - works fine on my laptop. 😕

**Comment by v-dobrev** [comment]
After some tweaks, I also got inconsistent behavior. It turned out the pressure, `p_gf`, is not initialized. Thanks to valgrind: it showed that there are uninitialized values, and they are allocated on line 263 in ex19p!

I'll push a fix - initialize the pressure to 0 - in a moment.

File: linalg/blockoperator.hpp [hunk 11]
```diff
@@ -74,7 +74,7 @@ class BlockOperator : public Operator
    //! Check if block (i,j) is a zero block
    int IsZeroBlock(int i, int j) const { return (op(i,j)==NULL) ? 1 : 0; }
    //! Return a reference to block i,j
-   Operator & GetBlock(int i, int j)
+   Operator & GetBlock(int i, int j) const
```
Comment (v-dobrev) [inline linalg/blockoperator.hpp]: This `const` does not seem to be necessary for building the library, examples, and miniapps. Is it really needed?

File: fem/nonlininteg.hpp [hunk 12]
```diff
@@ -72,6 +72,42 @@ class NonlinearFormIntegrator
 };
 
 
+class BlockNonlinearFormIntegrator
+{
+public:
+   /// Compute the local energy
+   virtual double GetElementEnergy(Array<const FiniteElement *>&el,
+                                   ElementTransformation &Tr,
+                                   Array<const Vector *>&elfun);
```
Comment (v-dobrev) [inline fem/nonlininteg.hpp]: It may be a bit more expressive to switch the `Array<...>` types to `const Array<...>`. What do you think?

File: fem/nonlininteg.hpp [hunk 13]
```diff
@@ -72,6 +72,42 @@ class NonlinearFormIntegrator
...
+
+   /// Perform the local action of the BlockNonlinearFormIntegrator
+   virtual void AssembleElementVector(Array<const FiniteElement *> &el,
+                                      ElementTransformation &Tr,
+                                      Array<Vector *> &elfun,
+                                      Array<Vector *> &elvec);
```
Comment (v-dobrev) [inline fem/nonlininteg.hpp]: The `elfun` arguments should probably be arrays of `const Vector *`.

File: linalg/blockoperator.hpp (hunk 11 above)
Comment (jamiebramwell) [reply linalg/blockoperator.hpp]: I think for some reason I had to add it when I hacked in the direct system solve. I'll remove it.

File: fem/nonlininteg.hpp (hunk 12 above)
Comment (jamiebramwell) [reply fem/nonlininteg.hpp]: That makes sense to me. I'll try it out and see if any issues pop up.

File: fem/nonlininteg.hpp (hunk 13 above)
Comment (jamiebramwell) [reply fem/nonlininteg.hpp]: Done. I had to add a new variable (`el_x_const`)[https://github.com/mfem/mfem/blob/7e6b0be6b7e28e835c9376bb18324bacd39d63ed/fem/nonlinearform.cpp#L455] to add the const declaration inside the template parameter. I thought this would be better than using `const_cast`, but let me know if you think otherwise.

**Comment by tzanio** [comment]
Awesome new feature and example @jamiebramwell!

glvis_s01 2

File: examples/ex19.cpp [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+#include <memory>
+#include <iostream>
+#include <fstream>
+
+using namespace std;
+using namespace mfem;
+
+// Custom block preconditioner for the Jacobian of the incompressible nonlinear
+// elasticity operator.
+class JacobianPreconditioner : public Solver
```
Comment (tzanio) [inline examples/ex19.cpp]: Document the form of the preconditioner?

Comment (jamiebramwell) [reply examples/ex19.cpp]: done.

File: fem/nonlininteg.hpp [hunk 15]
```diff
@@ -216,6 +252,35 @@ class HyperelasticNLFIntegrator : public NonlinearFormIntegrator
                                     const Vector &elfun, DenseMatrix &elmat);
 };
 
+
+class IncompressibleNeoHookeanIntegrator : public BlockNonlinearFormIntegrator
```
[the remark this answers is not in the document]
Comment (jamiebramwell) [reply fem/nonlininteg.hpp]: done.

**Comment by jamiebramwell** [comment]
I believe I have addressed all of the comments. Let me know if you have anything else you would like changed.

**Comment by tzanio** [comment]
Looks great, thanks @jamiebramwell!

**Comment by tzanio** [comment]
@jdahm and @acfisher -- can you take a look / approve?

File: fem/nonlininteg.cpp [hunk 16]
```diff
@@ -385,4 +439,230 @@ void HyperelasticNLFIntegrator::AssembleElementGrad(const FiniteElement &el,
...
+                 " has incorrect block finite element space size!");
+   }
+
+   int dof_u = el[0]->GetDof();
+   int dim = el[0]->GetDim();
+
+   DSh_u.SetSize(dof_u, dim);
+   J0i.SetSize(dim);
+   J1.SetSize(dim);
+   J.SetSize(dim);
```
Comment (acfisher) [inline fem/nonlininteg.cpp]: Do we want to support surface meshes embedded in 3D or is that a feature for another day?  If so spaceDim!=dim

File: fem/nonlininteg.cpp [hunk 17]
```diff
@@ -385,4 +439,230 @@ void HyperelasticNLFIntegrator::AssembleElementGrad(const FiniteElement &el,
...
+      el[1]->CalcShape(ip, Sh_p);
+      double pres = Sh_p * *elfun[1];
+      double mu = c_mu->Eval(Tr, ip);
+      double dJ = F.Det();
+      double dJ_FinvT_DS;
+
+      CalcInverseTranspose(F, FinvT);
+
+      // u,u block
+      for (int i_u = 0; i_u < dof_u; i_u++)
```
Comment (acfisher) [inline fem/nonlininteg.cpp]: It doesn't matter for integers these days so I wouldn't change it here, but its a good habit to use the ++i operator rather than the i++ operator for esoteric performance reasons.

Comment (jamiebramwell) [reply fem/nonlininteg.cpp]: Good to know! I always remember one is better, but I can never remember which is which.

File: fem/nonlininteg.cpp (hunk 16 above)
Comment (jamiebramwell) [reply fem/nonlininteg.cpp]: Manifold problems don't really make sense for these hyperelastic models as the 2D case only really makes sense if you consider it to be in plane strain. For the manifold case, you would want a full shell or plate model. I'll add a check to make sure it's not trying to be used in that way.
